lib/AST/Module.cpp

```diff
@@ -2572,6 +2572,38 @@ SourceFile::setImports(ArrayRef<AttributedImport<ImportedModule>> imports) {
   Imports = getASTContext().AllocateCopy(imports);
 }
 
+std::optional<AttributedImport<ImportedModule>>
+SourceFile::findImport(const ModuleDecl *module) const {
+  return evaluateOrDefault(getASTContext().evaluator,
+                           ImportDeclRequest{this, module}, std::nullopt);
+}
+
+std::optional<AttributedImport<ImportedModule>>
+ImportDeclRequest::evaluate(Evaluator &evaluator, const SourceFile *sf,
+                            const ModuleDecl *module) const {
+  auto &ctx = sf->getASTContext();
+  auto imports = sf->getImports();
+
+  // Look to see if the owning module was directly imported.
```

Comment:
There's an overload of ImportCache::getImportSet() that takes an ArrayRef<ImportedModule>. Maybe it can replace this whole thing?

Reply:
This was originally located in TypeCheckConcurrency, so there is a high  chance that it’s possible to dedupe…

Reply:
It looks like `ImportSet` is a flat set of imports, so the context of which top level import each transitive import comes from is lost. Correct me if I'm missing something but I don't see how that can be used to implement this request that needs to return the specific import from which a module is reachable.